### PR TITLE
Add an upgrade script for revpi nodes via npm

### PR DESCRIPTION
**Pull request by naruxde**
This change packages the appropriate node-red-contrib-revpi-nodes version into the noderedrevpinodes-server package. In the first step, this archive is only installed on the target system. In addition, an update script is provided that allows the user to update the existing NodeRED installation with the provided node version. The user rights must be observed.

### CODE REVIEW

File: debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes [hunk 1]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+#
+# SPDX-License-Identifier: GPL-2.0
+# Copyright: 2023 KUNBUS GmbH
```
Comment (tboehler1) [inline debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes]: This should be replaced with `SPDX-FileCopyrightText: 2023 KUNBUS GmbH`

File: debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes [hunk 2]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
```
Comment (LukasZech) [inline debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes:1]: Is there a convention wether to use

`#!/bin/sh`
or
`#!/bin/bash`
?

I could find both in multiple scripts in our repositories

File: debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+# As long as the node-red is installed in /home/pi/.node-red, we have to
+# upgrade the nodes in that location as user pi.
+
+NODE_RED_ROOT="/home/pi/.node-red"
+REVPI_NODES_VERSION="1.1.0"
+
+# Check node red installation directory
+if [ ! -d "${NODE_RED_ROOT}" ]; then
+    echo "NodeRED is not installed at '${NODE_RED_ROOT}'"
```
Comment (tboehler1) [inline debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes]: Redirect errors to `stderr`: `>&2`

File: debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes [hunk 4]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+#
+# SPDX-License-Identifier: GPL-2.0
```
Comment (tboehler1) [inline debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes]: SPDX License Identifier `GPL-2.0` is deprecated, use `GPL-2.0-only` or `GPL-2.0-or-later` instead.

https://spdx.org/licenses/GPL-2.0.html

File: debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+if [ "$(id -un)" == "pi" ]; then
+    # This will install/upgrade the revpi nodes of this package as pi user
+    /usr/bin/npm install \
+        --save --save-exact \
+        --prefix "${NODE_RED_ROOT}" \
+        "/usr/share/noderedrevpinodes-server/npm-install/${REVPI_NODES_VERSION}.tar.gz"
+
+elif [ "$(id -u)" -eq 0 ]; then
+    # As root we change the user to pi and execute this script again
+    /usr/bin/sudo -u pi "${0}"
```
Comment (tboehler1) [inline debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes]: I don't really like doing this as part of the script. Rather, I think it should be run by root like this in the first place if root is running this.
Feel free to disagree :p

File: debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+else
+    echo "This script must be executed as user pi or root."
```
Comment (tboehler1) [inline debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes]: Redirect to `stderr`

File: debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+if [ ! -d "${NODE_RED_ROOT}" ]; then
+    echo "NodeRED is not installed at '${NODE_RED_ROOT}'"
+    exit 1
+fi
+
+# Check user is root or pi, both can upgrade the nodes
+if [ "$(id -un)" == "pi" ]; then
+    # This will install/upgrade the revpi nodes of this package as pi user
+    /usr/bin/npm install \
+        --save --save-exact \
```
Comment (LukasZech) [inline debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes]: `--save-exact` has no effect for packages installed from a tarball as the version in the package.json will simply be set to the filename of the tarball

**Review by tboehler1 — CHANGES_REQUESTED** [review]

File: debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes (hunk 7 above)
Comment (LukasZech) [inline debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes]: `--save` is the default and its not necessary to provide it 
although imho it does not hurt here to make clear that this save to the package.json in the target dir

**Review by LukasZech — CHANGES_REQUESTED** [review]

File: debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes (hunk 2 above)
Comment (naruxde) [reply debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes:1]: On a debian system the /bin/sh is linked to the dash. The differences are probably in the range of functions, in dealing with variables, etc. However, the Bash is specified as Essential in Debian, which is defined as the minimal set of functionality that must be available and usable on the system at all times. Don't know 🤷

**Comment by LukasZech** [comment]
I don't quite understand the general approach of this.

If you just want to install a fixed version of the package, you could simply use
`npm install --save-exact node-red-contrib-revpi-nodes@1.1.0`

or if you want to install a version from the repository that was not yet release to npm (like v1.1.10) you can still install the tarball via url from there.

The problem with this approach is, that it changes the `/home/pi/.node-red/package.json` in a way, that it can never be updated though npm again as it replaces the version of the dependency with the path to the local tarball.


If you just want to enforce that a only specific version of a package is use on your system, simply install it as global dependency and remove the local dependency from `/home/pi/.node-red/package.json` alltogheter.

That would be a more proper approach than manipulating the dependencies of another package by installing a script in a different package from to different package manager to circumvent npm

`npm install --global node-red-contrib-revpi-nodes@1.1.0`

for the global package it could even make sense to use the included tarball

File: debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes (hunk 2 above)
Comment (tboehler1) [reply debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes:1]: My default is using POSIX sh and switching to bash if needed, but I don't know how others handle this.

I was wondering, in this case and generally, if we should use `#!/bin/bash` or `#!/usr/bin/env bash` for bash.

File: debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes (hunk 5 above)
Comment (iluminat23) [reply debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes]: > I don't really like doing this as part of the script. Rather, I think it should be run by root like this in the first place if root is running this. Feel free to disagree :p
As the intention is **not** running as root, your assumption is wrong.

But I would run this with `su` not `sudo`.

Comment (iluminat23) [reply debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes]: Or just bail if the wrong user is running the script.

**Comment by LukasZech** [comment]
Another approach would be to provide the package from the  `noderedrevpinodes-server` folder itself by including it in the source and link the dependency locally by using [npm link](https://docs.npmjs.com/cli/v9/commands/npm-link) which is  " handy for installing your own stuff" as the documentation quotes

File: debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes (hunk 5 above)
Comment (tboehler1) [reply debian/node-red-contrib-revpi-nodes/update-node-red-revpi-nodes]: > Or just bail if the wrong user is running the script.

This is what I mean. If the user `pi` should be running this then that user should be the only one calling it. root can achieve this with `su` if it really wants too, but this should be done *explicitly*.

**Comment by naruxde** [comment]
> Another approach would be to provide the package from the `noderedrevpinodes-server` folder itself by including it in the source and link the dependency locally by using [npm link](https://docs.npmjs.com/cli/v9/commands/npm-link) which is " handy for installing your own stuff" as the documentation quotes

I like that! Take a branch and show us how!

**Review by tboehler1 — APPROVED** [review]
The implementation is fine by me.

**Comment by LukasZech** [comment]
> > Another approach would be to provide the package from the `noderedrevpinodes-server` folder itself by including it in the source and link the dependency locally by using [npm link](https://docs.npmjs.com/cli/v9/commands/npm-link) which is " handy for installing your own stuff" as the documentation quotes
> 
> I like that! Take a branch and show us how!

@naruxde 
here you go

https://github.com/RevolutionPi/noderedrevpinodes-server/pull/9

**Comment by iluminat23** [comment]
This will ultimately kill any installation without internet access. Or am I mistaken?

**Comment by LukasZech** [comment]
including the node-red-contrib-revpi-nodes npm package in the noderedrevpinodes-server debian package makes the process completely offline.

the node-red-contrib-revpi-nodes npm package will then be updated with each new version of the noderedrevpinodes-server debian package

if the debian package is available offline then the node-red-contrib-revpi-nodes npm package is as well, as nothing is acutally installed via npm, at least using the `npm link` approach

installing the tarball via npm will install the dependencies listed in the package.json of the node-red-contrib-revpi-nodes npm package, although this is only because they are not declared correctly. If they were declared as peerDependency then nothing would need to be installed via npm at all

**Review by LukasZech — APPROVED** [review]

**Comment by LukasZech** [comment]
According to the documentation, it should not be necessary to install the nodes as npm package at all.
It seems to be sufficient if the files are present in the `nodesDir` folder
https://nodered.org/docs/user-guide/runtime/configuration
https://github.com/node-red/node-red/pull/3676


It should be enough to simply replace the `npm install tarball.tar.gz` with `untar tarball.tar.gz. /home/pi/.node-red/nodes`

**Comment by LukasZech** [comment]
just tested it and this works:
 ```

mkdir -p /home/pi/.node-red/nodes
tar -xf 1.1.0.tar.gz -C /home/pi/.node-red/nodes
```
after extracting the tar to this folder and restarting the node-red server the nodes are available in node-red without the need to use npm at all

**Comment by iluminat23** [comment]
OK, this needs to be done by the user. We will never add anything to a package which installs something to the home of the user.
May be an update script invoked by the user?

**Comment by naruxde** [comment]
But what about putting it to `/usr/lib/node_modules/`? Will this work for the node-red installation, too? Is that a kind of global library storage on a debian system? Like for example `/usr/lib/python3/dist-packages/` for python 3? @LukasZech

**Comment by naruxde** [comment]
All the solutions mentioned are wrong. We now need to focus on how to install the `node-red-contrib-revpi-nodes` in `/usr/lib/node_modules/node-red/node_modules`.

**Pull request closed without merging**